pkg/statistics/handle/usage/indexusage/collector.go

```diff
@@ -0,0 +1,225 @@
+}
+
+func (m indexUsage) update(tableID int64, indexID int64, sample Sample) {
+	id := GlobalIndexID{TableID: tableID, IndexID: indexID}
+	m.updateByKey(id, sample)
+}
+
+func (m indexUsage) merge(destMap indexUsage) {
+	for id := range destMap {
+		item := destMap[id]
+		m.updateByKey(id, item)
+	}
+}
+
+// Collector represents a data structure to record the index usage for the whole node
+type Collector struct {
+	collector collector.GlobalCollector[indexUsage]
+	mapper    indexUsage
+	sync.RWMutex
+
+	closed bool
+}
+
+// NewCollector create an index usage collector
+func NewCollector() *Collector {
+	iuc := &Collector{
+		mapper: make(indexUsage),
+	}
+	iuc.collector = collector.NewGlobalCollector[indexUsage](iuc.merge, indexUsageCollectorFlushTimeout, indexUsageCollectorChannelSize)
+
+	return iuc
+}
+
+// GetIndexUsage returns the index usage information
+func (c *Collector) GetIndexUsage(tableID int64, indexID int64) *Sample {
+	c.RLock()
+	defer c.RUnlock()
+
+	info, ok := c.mapper[GlobalIndexID{tableID, indexID}]
+	if !ok {
+		return nil
+	}
+	return &info
+}
+
+func (c *Collector) merge(delta indexUsage) {
+	c.Lock()
+	defer c.Unlock()
+
+	c.mapper.merge(delta)
+}
+
+// SpawnSessionCollector creates a new session collector attached to this global collector
+func (c *Collector) SpawnSessionCollector() *SessionIndexUsageCollector {
+	return &SessionIndexUsageCollector{
+		mapper:    make(indexUsage),
+		collector: c.collector.SpawnSession(),
+	}
+}
+
+// Close closes the background worker inside
+func (c *Collector) Close() {
+	c.collector.Close()
+}
+
+// GCIndexUsage will delete the usage information of non-existent indexes.
+// `tableMetaLookup` argument is represented as a function (but not `sessionctx.Context` or `infoschema.InfoSchema`) to
+// avoid depending on `sessionctx.Context`.
+func (c *Collector) GCIndexUsage(tableMetaLookup func(id int64) (*model.TableInfo, bool)) {
+	// it's possible to split `s.Mutex` into multiple mutex to avoid blocking the creation of session and sweepiing index
+	// However, as all these operations are infrequent, keeping a simpler mutex is enough.
+	c.Lock()
+	defer c.Unlock()
+	for k := range c.mapper {
+		tbl, ok := tableMetaLookup(k.TableID)
+		if !ok {
+			delete(c.mapper, k)
+			continue
+		}
+		foundIdx := false
+		for _, idx := range tbl.Indices {
+			if idx.ID == k.IndexID {
+				foundIdx = true
+				break
+			}
+		}
+		if !foundIdx {
+			delete(c.mapper, k)
+		}
+	}
+}
+
+// SessionIndexUsageCollector collects index usage per-session
+type SessionIndexUsageCollector struct {
+	mapper indexUsage
+
+	collector collector.SessionCollector[indexUsage]
+}
+
+// Update updates the mapper in SessionIndexUsageCollector
+func (s *SessionIndexUsageCollector) Update(tableID int64, indexID int64, sample Sample) {
+	s.mapper.update(tableID, indexID, sample)
+}
+
+// Report reports the mapper in `SessionIndexUsageCollector` to the global collector
+func (s *SessionIndexUsageCollector) Report() {
```

Comment:
This function is not called in this PR. I prepare to use it after (or before) each statement. When the collector for each statements (which is required to record `query count` for indexes) is reset, it'll also try to call the `Report`.